folding-schemes/src/frontend/noname/mod.rs

```diff
@@ -0,0 +1,228 @@
+        let (code, state_len, external_inputs_len) = params;
+        let compiled_circuit = compile_source_code::<BF>(&code).map_err(|_| {
+            Error::Other("Encountered an error while compiling a noname circuit".to_owned())
+        })?;
+        Ok(NoNameFCircuit {
+            state_len,
+            external_inputs_len,
+            circuit: compiled_circuit,
+            _f: PhantomData,
+        })
+    }
+
+    fn state_len(&self) -> usize {
+        self.state_len
+    }
+
+    fn external_inputs_len(&self) -> usize {
+        self.external_inputs_len
+    }
+
+    fn step_native(
+        &self,
+        _i: usize,
+        z_i: Vec<F>,
+        external_inputs: Vec<F>,
+    ) -> Result<Vec<F>, crate::Error> {
+        // compute witness using noname witness generator and outputs the next step
+        let external_inputs_json = match !external_inputs.is_empty() {
+            false => None,
+            true => {
+                let json =
+                    serde_json::to_string(&NoNameJSONExternalInputs::from(external_inputs.clone()))
+                        .map_err(|e| Error::JSONSerdeError(e.to_string()))?;
+                Some(json)
+            }
+        };
+
+        let ivc_inputs = serde_json::to_string(&NoNameJSONIVCInput::from(z_i.clone()))
+            .map_err(|e| Error::JSONSerdeError(e.to_string()))?;
+        let noname_inputs = NoNameIVCInputs::<F>::try_from((ivc_inputs, external_inputs_json))
+            .map_err(|_| SynthesisError::Unsatisfiable)?;
+        let (ivc_inputs, noname_external_inputs) = noname_inputs.try_into()?;
+        let noname_witness = self
+            .circuit
+            .generate_witness(ivc_inputs, noname_external_inputs)
+            .map_err(|e| Error::WitnessCalculationError(e.to_string()))?;
+
+        let mut assigned_z_i1 = vec![];
+        let z_i1_end_index = z_i.len() + 1;
+        for idx in 1..z_i1_end_index {
+            let value: BigUint = Into::into(noname_witness.witness[idx]);
+            let field_element = F::from(value);
+            assigned_z_i1.push(field_element);
+        }
+
+        Ok(assigned_z_i1)
+    }
+
+    fn generate_step_constraints(
+        &self,
+        cs: ark_relations::r1cs::ConstraintSystemRef<F>,
+        _i: usize,
+        z_i: Vec<ark_r1cs_std::fields::fp::FpVar<F>>,
+        external_inputs: Vec<ark_r1cs_std::fields::fp::FpVar<F>>,
+    ) -> Result<Vec<ark_r1cs_std::fields::fp::FpVar<F>>, ark_relations::r1cs::SynthesisError> {
+        let external_inputs_json = match !external_inputs.is_empty() {
+            false => None,
+            true => {
+                let json = serde_json::to_string(&NoNameJSONExternalInputs::from_fpvars(
+                    &external_inputs.clone(),
+                )?)
+                .map_err(|_| SynthesisError::Unsatisfiable)?;
+                Some(json)
+            }
+        };
+
+        let ivc_inputs = serde_json::to_string(&NoNameJSONIVCInput::from_fpvars(&z_i)?)
```

Comment:
There is a bunch of transformation that we need to do. The flow is: we receive the state and external inputs as `FpVars`, we turn those into `F` elements, those are converted into properly formatted string json inputs (using `serde_json::to_string`) and are finally turned into a data structure that is used as inputs for `generate_witness`.

Reply:
Thanks for the explanation! This comment was useful ^^ how would you see adding it as an inline comment in the code for future readers (including ourselves)?

Reply:
I have a question here.

Why would we recieve FpVars in the first place? Can't we get plain field elements directly and we generate the vars if needed too? Seems less complex.

Reply:
In sonobe, the `generate_step_constraints` method is used to compute the witness/trace of the `F` function, using the public state `z_i` and `external_inputs`. Both for design and security concerns (preventing users to allocate underconstrained witness values - cc @arnaucube, correct me if I'm wrong),`z_i` and `external_inputs` are allocated beforehand (when initializing the `AugmentedF` function). 
I agree that its kind of confusing (I asked about this to @arnaucube in the past!), but became convinced that's it's actually a nice way to do things! 
Still, if we wanted to get plain field elements, we will have to re-write the `FCircuit` trait and do a refactor of a few parts of the codebase.

Reply:
Ahh that's fair
If a refactor is needed let's ignore that for now. We can revisit this in the future then.